flink-tests/src/test/java/org/apache/flink/test/scheduling/ReactiveModeITCase.java

```diff
@@ -196,52 +192,91 @@ public void run(SourceContext<String> ctx) throws Exception {
 
-        private static final InstanceTracker instances = new InstanceTracker();
+        private volatile int instances = 0;
 
-        public static void expectInstances(int count) {
-            instances.expectInstances(count);
+        public void reportStoppedInstance() {
+            synchronized (lock) {
+                instances--;
+            }
         }
 
-        public static void waitForInstances() throws InterruptedException {
-            instances.waitForInstances();
+        public void reportNewInstance() {
+            synchronized (lock) {
+                instances++;
+                lock.notifyAll();
+            }
         }
 
-        @Override
-        public void open(Configuration parameters) throws Exception {
-            instances.reportNewInstance();
+        public void waitForScaleUpToParallelism(int parallelism) throws InterruptedException {
+            while (true) {
+                synchronized (lock) {
+                    if (instances == parallelism) {
+                        break;
+                    }
+                    lock.wait();
+                }
+            }
+        }
+
+        public void reset() {
+            synchronized (lock) {
+                instances = 0;
+            }
         }
     }
 
-    private static class InstanceTracker {
-        private final Object lock = new Object();
+    private static class FailOnParallelExecutionSource extends RichParallelSourceFunction<String> {
+        private volatile boolean running = true;
 
-        @GuardedBy("lock")
-        private CountDownLatch latch = new CountDownLatch(0);
+        private static final InstanceParallelismTracker tracker = new InstanceParallelismTracker();
 
-        public void reportNewInstance() {
-            synchronized (lock) {
-                if (latch.getCount() == 0) {
-                    throw new RuntimeException("Test error. More instances than expected.");
+        public static void waitForScaleUpToParallelism(int parallelism)
+                throws InterruptedException {
+            tracker.waitForScaleUpToParallelism(parallelism);
+        }
+
+        public static void resetParallelismTracker() {
+            tracker.reset();
+        }
+
+        @Override
+        public void open(Configuration parameters) throws Exception {
+            if (getRuntimeContext().getNumberOfParallelSubtasks() > 1) {
+                throw new IllegalStateException(
+                        "This is not supposed to be executed in parallel, despite extending the right base class.");
+            }
+        }
+
+        @Override
+        public void run(SourceContext<String> ctx) throws Exception {
+            tracker.reportNewInstance();
+            while (running) {
+                synchronized (ctx.getCheckpointLock()) {
+                    ctx.collect("test");
                 }
-                latch.countDown();
+                Thread.sleep(100);
```

Comment:
Can this be decreased to `10` or `20`? Otherwise the test will run at least 100ms.